nav2_util/include/nav2_util/validate_messages.hpp

```diff
@@ -0,0 +1,106 @@
+//  Check recieved message is safe or not for the nav2-system
+//  For each msg-type known in nav2, we could check it as following:
+//  if(!validateMsg()) RCLCPP_ERROR(,"malformed msg. Rejecting.")
+//
+//  Workflow of validateMsg():
+//     if here's a sub-msg-type in the recieved msg,
+//        the content of sub-msg would be checked as sub-msg-type
+//     then, check the whole recieved msg.
+//
+//  Following conditions are involved in check:
+//     1> Value Check: to avoid damaged value like like `nan`, `INF`
+//     2> Logic Check: to avoid value with bad logic, like the size of `map` should be equal to `height*width`
+//     3> Any other needed condition could be joint here in future
+
+#ifndef validate_message_HPP
+#define validate_message_HPP
+
+
+#include "nav_msgs/msg/occupancy_grid.hpp"
+#include "nav_msgs/msg/odometry.hpp"
+
+
+namespace nav2_util
+{
+
+const int NSEC_PER_SEC = 1e9;  // 1 second = 1e9 nanosecond
+
+bool validateMsg(const builtin_interfaces::msg::Time_<std::allocator<void>> & msg)
+{
+  // check item
```

Comment:
Rather, `if (msg.sec ==0 && msg.nanosec == 0)`, else it should be valid.

Nanosec is a `uint32`, so it cannot be negative. Seconds can be, but I think that actually is valid in some odd cases

Reply:
> which default is (0,0)

Since I noticed it, so I think `if (msg.sec ==0 && msg.nanosec == 0)` it still should be valid ?

> Nanosec is a uint32, so it cannot be negative. Seconds can be

well, if negative seconds is allowed, I'm going to change this function as following ? 

```cpp
bool validateMsg(const builtin_interfaces::msg::Time & msg)
{
    // check item 
    // negative second is allowed 
    if (msg.nanosec >= NSEC_PER_SEC) return false;  // invalid nanosec-stamp
    return true; 
}


```

Reply:
When would `msg.sec ==0 && msg.nanosec == 0` be valid?

Reply:
I catched a map message by `ros2 topic echo /map` during launching nav2, and its header is:

```yaml
# by: ros2 topic echo /map
header:
  stamp:
    sec: 0
    nanosec: 0
  frame_id: map
...
```

Reply:
I'd argue that's invalid and we should fix that as part of this PR :-) You've uncovered a problem by checking the validity of messages, that's good!

Reply:
I see. It seems that we need to add a timestamp-mechanism to the publisher of the map. I'll update this PR over the weekend or possibly later, trying to address all of today's issues in one go. :-)

Reply:
well , I had a try to find how time-stamp of a map-message is set by `nav2_map_server` as following:

```cpp
  // Since loadMapFromFile() does not belong to any node, publishing in a system time.
  rclcpp::Clock clock(RCL_SYSTEM_TIME);
  msg.info.map_load_time = clock.now();
  msg.header.frame_id = "map";
  msg.header.stamp = clock.now();
```

HOWEVER, I `ros2 topic echo /map` during launching nav2 for many times, while the time-stamp of message I got is still 
```yaml
# by: ros2 topic echo /map
header:
  stamp:
    sec: 0
    nanosec: 0
  frame_id: map
...
```

It's odd to me, I'd have more check in code later.

Reply:
It' so strange that I can't understand what happened. The time-stamp of message I catch by command is always 
```yaml
sec:0
nanosec:0
```
Fortunately, this issue does not affect the usage and operation of navigation2. Perhaps there is some mechanism in place that I am not familiar with, ensuring this aspect remains unaffected.

Reply:
I don't understand. I assume there needs to be some updates to missing timestamps like you mentioned? That + running all the tests to make sure nothing else breaks + fixing the linting issues is the last before merging